torch/csrc/autograd/engine.cpp

```diff
@@ -1155,31 +1171,38 @@ void GraphTask::init_to_execute(Node& graph_root, const edge_list& outputs, bool
       return nullptr;
     }
   };
+
+  auto nodeShouldExecute = [this](Node *fn) {
+    auto it = exec_info_.find(fn);
+    return it != exec_info_.end() && it->second.should_execute();
+  };
+
   std::vector<Frame> stack;
   std::unordered_set<Node*> seen;
-  for (const auto & input : graph_root.next_edges()) {
-    if (!input.function.get()) continue;
-    if (seen.count(input.function.get()) > 0) continue;
-    stack.emplace_back(input.function.get());
-    while (!stack.empty()) {
-      auto &frame = stack.back();
-      if (Node *next_fn = frame.get_next_fn()) {
-        if (/* bool unseen = */ seen.emplace(next_fn).second) {
-          stack.emplace_back(next_fn);
-          continue; // recurse
-        }
-      } else {
-        // NB: if we were using real recursion we could have saved some lookups
-        // using a return value from recursive call. It would make this manually unrolled
-        // version a lot more complicated, so I skipped that.
-        const auto & next_edges = frame.fn_->next_edges();
-        const bool needed = std::any_of(
-            next_edges.begin(), next_edges.end(), [&](const Edge& edge) {
-              auto it = exec_info_.find(edge.function.get());
-              return it != exec_info_.end() && it->second.should_execute();
-            });
-        exec_info_[frame.fn_].needed_ |= needed;
-        stack.pop_back();
+  stack.emplace_back(&graph_root);
+  exec_info_.emplace(stack.back().fn_, ExecInfo());
+
+  while (!stack.empty()) {
+    auto &frame = stack.back();
+    const auto fn = frame.fn_;
+
+    Node *child_fn = nullptr;
+    while((child_fn = frame.get_next_fn()) && !seen.emplace(child_fn).second) {
+      // (1) next child exists AND has already been seen
+      if (nodeShouldExecute(child_fn)) {
+        exec_info_[fn].needed_ = true;
+      }
+    }
+
+    if (child_fn) {
+      // (2) next child exists but has not been seen
+      stack.emplace_back(child_fn);
+    } else {
+      // (3) no next child exists for `fn` means its `needed` has already been
+      // finalized. pop stack and update parent
+      stack.pop_back();
+      if (!stack.empty() && nodeShouldExecute(fn)) {
```

Comment:
Swapping those two conditions may lead to better performance , because `!stack.empty()` is false only for the root node. 
```
if (nodeShouldExecute(fn) && !stack.empty()) {
```
`!stack.empty()` condition can be eliminated if a sentinel node is added as the parent of the root, though that would make the code less readable.

Reply:
good catch. The sentinel node is an interesting idea, but yeah lets not over optimize here at the cost of code readability